src/streamlink/plugins/youtube.py

```diff
@@ -364,6 +412,68 @@ def _get_streams(self):
                               "try youtube-dl instead")
 
         return streams
+    
+    def handle_403(self, url, query):
+        res = http.get(url, params=query, raise_for_status=False)
+        if res.status_code == codes.ok:
+            return res
+        else:
+            while res.status_code == codes.forbidden:
```

Comment:
This could loop for quite a while, it might be nice to have some retries here and a graceful fallback.

```python
def handle_403(self, url, query, retries=5):
    for _ in range(retries):
        res = http.get(url, params=query, raise_for_status=False)
        if res.status_code == codes.ok:
            return res
        elif res.status_code == codes.forbidden:
            self.logger.debug("YouTube says \"Forbidden\". Retrying api call to {0} in 1 second", url)
            sleep (1)
        else:
            # other cases? break? retry?
```

Then in `set_title_info` you check to see if it returns `None`, log a warning saying that the title information was not available at this time, and continue on.

Reply:
I would just set it to **None** on the first fail instead of this

Reply:
That might be best, but sometimes the youtube api key is just totally broken. Doesn't matter how long I backed off, (1 second vs 10 seconds, doesn't matter because other users are using the same API key). It just "wouldn't work" until I spammed the call enough times with handle_403. It gets very hard to be lucky enough times in a row to successfully get {title} {author} and {category} all at the same time, since each one is a separate api call.

Reply:
Maybe Streamlink needs more api keys to cycle through if we get a 403 error? I don't know much about their API or what it costs.

Reply:
Maybe @back-to is right. We discussed adding more API keys but decided against it. We did change the YouTube plugin to avoid the main use of the API (searching, which costs 100 api points, I think lookups are cheap 1 or 2 points), but most people are not using that version yet. But, we could change the API key, so that people using the new version are not penalised.

Reply:
Changing the API key sounds like a good idea. Was it a ToS issue that prevented adding a rotation of API keys?

Reply:
No one wanted to sign up the keys, you need to give a phone number iirc. Plus, it wasn’t needed after the updated plugin. If people upgrade to the latest version (I think it was merged for 0.13.0) they shouldn’t have any more 403 issues. For these API calls I think they use leas quota so there would need to be 30x as many users to cause the same problem (assuming we change the key).

Reply:
I don’t think we want to go the API key rotation route, as I said before it shouldn’t be an issue any more. We can open a new issue to discuss changing the API key for YouTube.

Reply:
I'll just change handle_403 to not retry as @back-to suggested. Will be renamed to handle other errors as you suggested.